File: client/ios/CodemagicPatchStorage.swift
import Foundation
import Darwin

final class CodemagicPatchStorage {
  static let shared = CodemagicPatchStorage()

  let root: URL

  private init() {
    let documents = FileManager.default
      .urls(for: .documentDirectory, in: .userDomainMask).first!
    self.root = documents.appendingPathComponent("codemagic-patch", isDirectory: true)
    try? FileManager.default.createDirectory(
      at: root.appendingPathComponent("state", isDirectory: true),
      withIntermediateDirectories: true
    )
  }

  // MARK: - Paths

  func url(_ relativePath: String) -> URL {
    root.appendingPathComponent(relativePath)
  }

  func packageContentsDir(_ packageHash: String) -> URL {
    root.appendingPathComponent("packages/\(packageHash)/contents", isDirectory: true)
  }

  // MARK: - Reads

  func readJson(_ relativePath: String) -> [String: Any]? {
    guard let data = try? Data(contentsOf: url(relativePath)) else { return nil }
    return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
  }

  func readState() -> CodemagicPatchState {
    guard let data = try? Data(contentsOf: url("state/state.json")),
          let state = try? JSONDecoder().decode(CodemagicPatchState.self, from: data) else {
      return CodemagicPatchState()
    }
    return state.sanitized()
  }

  func writeState(_ state: CodemagicPatchState) throws {
    let data = try JSONEncoder().encode(state)
    try writeBytes("state/state.json", data)
  }

  func mutateState(_ mutator: (inout CodemagicPatchState) throws -> Void) throws {
    var state = readState()
    try mutator(&state)
    try writeState(state)
  }

  func packageMetadata(_ packageHash: String) -> [String: Any]? {
    guard CodemagicPatchStorage.isSafePackageHash(packageHash) else { return nil }
    return readJson("packages/\(packageHash)/update.json")
  }

  func metadataMatchesBinary(packageHash: String, binaryVersion: String) -> Bool {
    guard CodemagicPatchStorage.isSafePackageHash(packageHash),
          let metadata = readJson("packages/\(packageHash)/update.json") else {
      return false
    }
    return metadata["package_hash"] as? String == packageHash &&
      metadata["binary_version"] as? String == binaryVersion
  }

  // MARK: - Writes

  func writeJson(_ relativePath: String, _ value: [String: Any]) throws {
    let data = try JSONSerialization.data(withJSONObject: value)
    try writeBytes(relativePath, data)
  }

  func writeText(_ relativePath: String, _ value: String) throws {
    try writeBytes(relativePath, Data(value.utf8))
  }

  func writeBytes(_ relativePath: String, _ data: Data) throws {
    try durableWriteData(data, to: url(relativePath))
  }

  func durableWriteData(_ data: Data, to url: URL) throws {
    let directory = url.deletingLastPathComponent()
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)

    let tmp = directory.appendingPathComponent(".\(url.lastPathComponent).\(UUID().uuidString).tmp")
    let fd = open(tmp.path, O_WRONLY | O_CREAT | O_TRUNC, mode_t(S_IRUSR | S_IWUSR))
    guard fd >= 0 else {
      throw CodemagicPatchStorage.posixError("open failed for \(tmp.lastPathComponent)")
    }

    var fdOpen = true
    do {
      try data.withUnsafeBytes { rawBuffer in
        guard let baseAddress = rawBuffer.baseAddress else { return }
        var written = 0
        while written < rawBuffer.count {
          let result = Darwin.write(
            fd,
            baseAddress.advanced(by: written),
            rawBuffer.count - written
          )
          guard result >= 0 else {
            throw CodemagicPatchStorage.posixError("write failed for \(tmp.lastPathComponent)")
          }
          written += result
        }
      }
      guard fsync(fd) == 0 else {
        throw CodemagicPatchStorage.posixError("fsync failed for \(tmp.lastPathComponent)")
      }
      let closeResult = close(fd)
      fdOpen = false
      guard closeResult == 0 else {
        throw CodemagicPatchStorage.posixError("close failed for \(tmp.lastPathComponent)")
      }
      guard rename(tmp.path, url.path) == 0 else {
        throw CodemagicPatchStorage.posixError("rename failed for \(url.lastPathComponent)")
      }
      fsyncDirectory(directory)
    } catch {
      if fdOpen {
        _ = close(fd)
      }
      try? FileManager.default.removeItem(at: tmp)
      throw error
    }
  }

  // MARK: - Filesystem maintenance

  func fsyncDirectory(_ directory: URL) {
    let fd = open(directory.path, O_RDONLY)
    guard fd >= 0 else { return }
    _ = fsync(fd)
    _ = close(fd)
  }

  func fsyncTree(_ url: URL) {
    var isDirectory: ObjCBool = false
    guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory) else {
      return
    }

    if isDirectory.boolValue {
      let children = (try? FileManager.default.contentsOfDirectory(
        at: url,
        includingPropertiesForKeys: nil,
        options: []
      )) ?? []
      children.forEach(fsyncTree)
      fsyncDirectory(url)
      return
    }

    let fd = open(url.path, O_RDONLY)
    guard fd >= 0 else { return }
    _ = fsync(fd)
    _ = close(fd)
  }

  func removeItemDurably(at url: URL) {
    let directory = url.deletingLastPathComponent()
    try? FileManager.default.removeItem(at: url)
    fsyncDirectory(directory)
  }

  func removeRelative(_ relativePath: String) {
    removeItemDurably(at: url(relativePath))
  }

  // MARK: - Events

  func eventFileURLs() -> [URL] {
    let dir = root.appendingPathComponent("events", isDirectory: true)
    let files = (try? FileManager.default.contentsOfDirectory(
      at: dir,
      includingPropertiesForKeys: [.contentModificationDateKey],
      options: [.skipsHiddenFiles]
    )) ?? []
    return files
      .filter { $0.pathExtension == "json" }
      .sorted { left, right in
        let leftDate = (try? left.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
        let rightDate = (try? right.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
        return leftDate < rightDate
      }
  }

  func enforceEventQueueCap(maxEvents: Int = 100) {
    let files = eventFileURLs()
    let overflow = files.count - maxEvents
    if overflow > 0 {
      files.prefix(overflow).forEach { try? FileManager.default.removeItem(at: $0) }
    }
  }

  // MARK: - Device id

  func getOrCreateDeviceId() -> String {
    let supportDir = FileManager.default
      .urls(for: .applicationSupportDirectory, in: .userDomainMask)
      .first!
      .appendingPathComponent("codemagic-patch", isDirectory: true)
    let deviceIdFile = supportDir.appendingPathComponent("device_id")

    if let existing = try? String(contentsOf: deviceIdFile, encoding: .utf8)
      .trimmingCharacters(in: .whitespacesAndNewlines),
       !existing.isEmpty {
      return existing
    }

    let defaults = UserDefaults.standard
    if let existing = defaults.string(forKey: "codemagic-patch.device_id"), !existing.isEmpty {
      try? FileManager.default.createDirectory(at: supportDir, withIntermediateDirectories: true)
      try? existing.data(using: .utf8)?.write(to: deviceIdFile, options: .atomic)
      return existing
    }

    let generated = UUID().uuidString
    try? FileManager.default.createDirectory(at: supportDir, withIntermediateDirectories: true)
    try? generated.data(using: .utf8)?.write(to: deviceIdFile, options: .atomic)
    defaults.set(generated, forKey: "codemagic-patch.device_id")
    return generated
  }

  // MARK: - Static helpers

  static func isSafePackageHash(_ packageHash: String) -> Bool {
    packageHash.range(of: #"^[a-f0-9]{64}$"#, options: .regularExpression) != nil
  }

  static func posixError(_ message: String) -> NSError {
    NSError(
      domain: NSPOSIXErrorDomain,
      code: Int(errno),
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }
}

File: client/android/src/main/java/io/codemagic/patch/CodemagicPatchStorage.kt
package io.codemagic.patch

import android.system.Os
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream
import java.util.UUID

internal class CodemagicPatchStorage(
    private val context: android.content.Context
) {
  val root: File = File(context.filesDir, "codemagic-patch")

  private val stateDir = File(root, "state")

  init {
    root.mkdirs()
    stateDir.mkdirs()
  }

  fun readJson(relativePath: String): JSONObject? {
    val file = File(root, relativePath)
    if (!file.isFile) return null
    return try {
      JSONObject(file.readText(Charsets.UTF_8))
    } catch (_: Exception) {
      null
    }
  }

  fun readText(relativePath: String): String? {
    val file = File(root, relativePath)
    return if (file.isFile) file.readText(Charsets.UTF_8) else null
  }

  fun listFiles(relativePath: String): List<File> {
    val dir = File(root, relativePath)
    return dir.listFiles()?.filter { it.isFile }?.toList() ?: emptyList()
  }

  fun writeJson(relativePath: String, json: JSONObject) {
    durableWrite(relativePath, json.toString().toByteArray(Charsets.UTF_8))
  }

  fun writeText(relativePath: String, value: String) {
    durableWrite(relativePath, value.toByteArray(Charsets.UTF_8))
  }

  fun writeBytes(relativePath: String, bytes: ByteArray) {
    durableWrite(relativePath, bytes)
  }

  fun writeStream(relativePath: String, input: InputStream, onProgress: ((Long) -> Unit)? = null): Long {
    val target = File(root, relativePath)
    return durableWrite(target) { output ->
      val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
      var written = 0L
      while (true) {
        val read = input.read(buffer)
        if (read < 0) break
        output.write(buffer, 0, read)
        written += read.toLong()
        onProgress?.invoke(written)
      }
      written
    }
  }

  fun readState(): CodemagicPatchState =
      CodemagicPatchState.fromJson(readJson("state/state.json") ?: JSONObject(), ::isSafePackageHash)

  fun writeState(state: CodemagicPatchState) {
    writeJson("state/state.json", state.toJson())
  }

  fun mutateState(mutator: (CodemagicPatchState) -> Unit) {
    val state = readState()
    mutator(state)
    writeState(state)
  }

  private fun durableWrite(relativePath: String, bytes: ByteArray) {
    val target = File(root, relativePath)
    durableWrite(target) { output ->
      output.write(bytes)
      Unit
    }
  }

  private fun <T> durableWrite(target: File, write: (FileOutputStream) -> T): T {
    val parent = target.parentFile ?: error("target has no parent")
    parent.mkdirs()
    val tmp = File(parent, ".${target.name}.${UUID.randomUUID()}.tmp")
    val result =
      try {
        FileOutputStream(tmp).use { output ->
          val value = write(output)
          output.fd.sync()
          value
        }
      } catch (error: Exception) {
        tmp.delete()
        throw error
      }
    try {
      Os.rename(tmp.absolutePath, target.absolutePath)
    } catch (error: Exception) {
      tmp.delete()
      throw error
    }
    fsyncDirectory(parent)
    return result
  }

  fun delete(relativePath: String) {
    val target = File(root, relativePath)
    val parent = target.parentFile
    target.deleteRecursively()
    if (parent != null) {
      fsyncDirectory(parent)
    }
  }

  fun fsyncDirectory(directory: File) {
    try {
      FileInputStream(directory).use { input ->
        input.fd.sync()
      }
    } catch (_: Exception) {
      // Directory fsync is best-effort across Android API levels/filesystems.
    }
  }

  fun fsyncTree(file: File) {
    if (file.isDirectory) {
      file.listFiles()?.forEach { fsyncTree(it) }
      fsyncDirectory(file)
      return
    }
    if (file.isFile) {
      FileInputStream(file).use { input ->
        input.fd.sync()
      }
    }
  }

  fun packageMetadata(packageHash: String): JSONObject? =
      if (isSafePackageHash(packageHash)) readJson("packages/$packageHash/update.json") else null

  fun packageContentsDir(packageHash: String): File =
      File(root, "packages/$packageHash/contents")

  fun metadataMatchesBinary(packageHash: String, binaryVersion: String): Boolean {
    if (!isSafePackageHash(packageHash)) return false
    val metadata = readJson("packages/$packageHash/update.json") ?: return false
    return metadata.optString("package_hash") == packageHash &&
        metadata.optString("binary_version") == binaryVersion
  }

  fun enforceEventQueueCap(maxEvents: Int = 100) {
    val committed = listFiles("events")
      .filter { it.name.endsWith(".json") }
      .sortedBy { it.lastModified() }
    val overflow = committed.size - maxEvents
    if (overflow > 0) {
      committed.take(overflow).forEach { it.delete() }
    }
  }

  fun isSafePackageHash(packageHash: String): Boolean =
      packageHash.isNotBlank() &&
          Regex("^[a-f0-9]{64}$").matches(packageHash)

  fun getDeviceId(): String {
    val prefs = context.getSharedPreferences("codemagic-patch", android.content.Context.MODE_PRIVATE)
    val existing = prefs.getString("device_id", null)
    if (!existing.isNullOrBlank()) return existing
    val generated = UUID.randomUUID().toString()
    prefs.edit().putString("device_id", generated).apply()
    return generated
  }

}
